clang/lib/Sema/SemaDecl.cpp

```diff
@@ -6107,6 +6109,44 @@ static bool isFromSystemHeader(SourceManager &SM, const Decl *D) {
          SM.isInSystemMacro(D->getLocation());
 }
 
+constexpr unsigned countCPlusPlusKeywords() {
+  unsigned Ret = 0;
+#define MODULES_KEYWORD(NAME)
+#define KEYWORD(NAME, FLAGS) ++Ret;
+#define CXX_KEYWORD_OPERATOR(NAME, TOK) ++Ret;
+#include "clang/Basic/TokenKinds.def"
+  return Ret;
+}
+
+static bool isKeywordInCPlusPlus(const Sema &S, const IdentifierInfo *II) {
+  if (!II)
+    return false;
+
+  // Build a static map of identifiers for all of the keywords in C++ that are
+  // not keywords in C. This allows us to do pointer comparisons instead of
+  // string comparisons when deciding whether the given identifier is a keyword
+  // or not. Note, this treats all keywords as being enabled, regardless of the
+  // setting of other language options. It intentionally disables the modules
+  // keywords because those are conditional keywords, so may be safe to use.
+  static auto Keywords = [&S] {
+    std::array<uintptr_t, countCPlusPlusKeywords()> Ret;
+    unsigned Idx = 0;
+#define MODULES_KEYWORD(NAME)
+#define KEYWORD(NAME, FLAGS)                                                   \
+  Ret[Idx++] = reinterpret_cast<uint64_t>(                                     \
+      &S.getPreprocessor().getIdentifierTable().get(#NAME));
+#define CXX_KEYWORD_OPERATOR(NAME, TOK)                                        \
+  Ret[Idx++] = reinterpret_cast<uint64_t>(                                     \
+      &S.getPreprocessor().getIdentifierTable().get(#NAME));
+#include "clang/Basic/TokenKinds.def"
+    assert(Idx == Ret.size() && "expected to fill every member!");
+    llvm::sort(Ret);
+    return Ret;
+  }();
+
+  return llvm::binary_search(Keywords, reinterpret_cast<uintptr_t>(II));
```

Comment:
A binary search for each identifier still seems excessively expensive. Can we instead just store a flag on the `IdentifierInfo` that we set up when building the initial table? That's what we do in other similar cases (eg, `IsExtension` / `IsFutureCompatKeyword`); it doesn't seem like we need to invent something different here.

Reply:
Based on our count, this leaves us with ~6 compares at worst, and only when the flag was enabled.  According to benchmarking, it is sub-0.5% perf-hit when the warning is enabled, so expense didn't really show up as overly measurable I think?

As far as the flag on IdentifierInfo... I'm amazed to find out we have bits left to do so!  Apparently we have 22 bits left according to the comment/my count confirming it.

I'd considered suggesting it at one point but apparently miscounted us having bits left over (an off-by-10 count is.... something), and didn't want every IdentifierInfo to pay the cost.  

I think this is going to be the best way forward here.  A slight startup cost, to calculate these, but makes this a way easier patch to swallow.

Reply:
Good suggestion, I've gone ahead with those changes.